src/main/java/org/jabref/logic/texparser/DefaultTexParser.java

```diff
@@ -0,0 +1,181 @@
+    }
+
+    @Override
+    public TexParserResult parse(List<Path> texFiles) {
+        return parseTexFiles(texFiles);
+    }
+
+    private TexParserResult parseTexFiles(List<Path> texFiles) {
+        TexParserResult result = new TexParserResult(masterDatabase);
+        int fileIndex = 0;
+
+        while (fileIndex < texFiles.size()) {
+            Path file = texFiles.get(fileIndex++);
+
+            try (LineNumberReader lnr = new LineNumberReader(Files.newBufferedReader(file))) {
+                for (String line = lnr.readLine(); line != null; line = lnr.readLine()) {
+                    // Skip comment lines.
+                    if (line.startsWith("%")) {
+                        continue;
+                    }
+
+                    matchCitation(result, file, lnr.getLineNumber(), line);
+                    matchNestedFile(result, file, texFiles, line);
+                }
+            } catch (IOException e) {
+                LOGGER.warn("Error opening the TEX file", e);
+            }
+        }
+
+        resolveTags(result);
+        return result;
+    }
+
+    /*
+     * Find cites along a specific line and store them.
+     */
+    private void matchCitation(TexParserResult result, Path file, int lineNumber, String line) {
+        Matcher citeMatch = Pattern.compile(CITE_REGEX).matcher(line);
+
+        while (citeMatch.find()) {
+            String[] keys = citeMatch.group(1).split(",");
+
+            for (String key : keys) {
+                addKey(result, key.trim(), new Citation(file, lineNumber, citeMatch.start(), citeMatch.end(), line));
+            }
+        }
+    }
+
+    private void matchNestedFile(TexParserResult result, Path file, List<Path> fileList, String line) {
+        Matcher includeMatch = Pattern.compile(INCLUDE_REGEX).matcher(line);
+
+        while (includeMatch.find()) {
+            String[] includes = includeMatch.group(1).split(",");
+
+            for (String include : includes) {
+                Path inputFile = (file.getParent() != null)
+                        ? file.getParent().resolve(include)
+                        : Paths.get(include);
+
+                if (!fileList.contains(inputFile)) {
+                    fileList.add(inputFile);
+                    result.increaseNestedFilesCounter();
+                }
+            }
+        }
+    }
+
+    /*
+     * Add a citation to the uniqueKeys map.
+     */
+    private void addKey(TexParserResult result, String key, Citation citation) {
+        Map<String, List<Citation>> uniqueKeys = result.getUniqueKeys();
+
+        if (!uniqueKeys.containsKey(key)) {
+            uniqueKeys.put(key, new ArrayList<>());
+        }
+
+        if (!uniqueKeys.get(key).contains(citation)) {
+            uniqueKeys.get(key).add(citation);
+        }
+    }
+
+    /*
+     * Look for an equivalent BibTeX entry within the reference database for all keys inside of the TEX files.
+     */
+    private void resolveTags(TexParserResult result) {
```

Comment:
Personally, I would split parsing the latex file and cross-referencing it with a given `BibDatabase` (it is a bit more flexible and in the spirit of "each class does only one job")

Reply:
All database-related methods are now in another class, but I still have to make some changes within the following days.